client/landing/gutenboarding/components/signup-form/index.tsx

```diff
@@ -48,6 +49,32 @@ const SignupForm = ( { onRequestClose }: Props ) => {
 		} );
 	}, [] );
 
+	useEffect( () => {
+		if ( newUserError ) {
+			let newErrorMessage: string | undefined;
+			switch ( newUserError.error ) {
+				case 'already_taken':
+				case 'already_active':
+				case 'email_exists':
+					newErrorMessage = __( 'An account with this email address already exists.' );
+					break;
+				case 'password_invalid':
+					newErrorMessage = newUserError.message;
+					break;
+				default:
+					newErrorMessage = __(
+						'Sorry, something went wrong when trying to create your account. Please try again.'
+					);
+					break;
+			}
+			setErrorMessage( newErrorMessage );
+			recordOnboardingError( {
+				step: 'account_creation',
+				error: newUserError.error || 'signup_form_new_user_error',
+			} );
+		}
+	}, [ newUserError ] );
```

Comment:
@p-jackson I've moved the errorMessage and tracking call into a useEffect block so that we set both the message and tracking only when the error changes.

Reply:
This does the trick, but I feel like we've been overusing `useEffect()` in gutenboarding in general. And adding side-effects to prop changes makes things trickier to follow (imo). We're also doubling up on the source of truth for where the error state lives: we have one that lives in Redux, on that lives in component state.

If we really want the event to be sent every time `newUserError` changes (not just when calling `createAccount()` fails) then maybe we should send it in the data layer.

In `stores/onboard/actions.ts`; `createSite()` is an example of an action that does a gutenboarding specific work before calling the generic version of `createSite()`. Maybe we could do something similar with firing gutenboarding analytics?

Reply:
I believe it will be inevitable to have analytics tied to both operations over the network and user interactions.

Absolutely, having some way to tie analytics into data-stores side effects is 💯. This sounds like a case where we should have analytics like:

- UI (from a component): User clicks signup
- Data (from data-stores analytics control): Signup sent
- Data (from data-stores analytics control): Signup success/fail
- UI (from a component): …?